File: product-microservice/api/keyboard_api.go
package api

import (
	"product/dto"
	"product/dto/filter"
	"product/mapper"
	"product/service"
	"strconv"

	"net/http"

	"github.com/gin-gonic/gin"
)


type KeyboardAPI struct {
	IKeyboardService service.IKeyboardService
}

func NewKeyboardAPI(keyboardService service.IKeyboardService) KeyboardAPI {
	return KeyboardAPI{IKeyboardService: keyboardService}
}

func (keyboardApi *KeyboardAPI) GetAll(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }
	
	keyboards := keyboardApi.IKeyboardService.GetAll(page, pageSize)
	c.JSON(http.StatusOK, mapper.ToKeyboardDTOs(keyboards))
}

func (keyboardApi *KeyboardAPI) GetNumberOfRecords(c *gin.Context) {
	numberOfRecords := keyboardApi.IKeyboardService.GetNumberOfRecords()
	c.JSON(http.StatusOK, numberOfRecords)
}

func (keyboardApi *KeyboardAPI) GetByID(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	keyboard, err := keyboardApi.IKeyboardService.GetById(id)
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToKeyboardDTO(keyboard))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (keyboardApi *KeyboardAPI) SearchByName(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }

	keyboards, err := keyboardApi.IKeyboardService.SearchByName(page, pageSize, c.Query("name"))
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToKeyboardDTOs(keyboards))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (keyboardApi *KeyboardAPI) GetNumberOfRecordsSearch(c *gin.Context) {
	numberOfRecords := keyboardApi.IKeyboardService.GetNumberOfRecordsSearch(c.Query("name"))
	c.JSON(http.StatusOK, numberOfRecords)
}

func (keyboardApi *KeyboardAPI) Filter(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }

	var filter filter.KeyboardFilter
	err = c.BindJSON(&filter)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	keyboards, err := keyboardApi.IKeyboardService.Filter(page, pageSize, filter)
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToKeyboardDTOs(keyboards))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (keyboardApi *KeyboardAPI) GetNumberOfRecordsFilter(c *gin.Context) {
	var filter filter.KeyboardFilter
	err := c.BindJSON(&filter)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	numberOfRecords := keyboardApi.IKeyboardService.GetNumberOfRecordsFilter(filter)
	c.JSON(http.StatusOK, numberOfRecords)
}

func (keyboardApi *KeyboardAPI) GetManufacturers(c *gin.Context) {
	manufacturers := keyboardApi.IKeyboardService.GetManufacturers()
	c.JSON(http.StatusOK, manufacturers)
}

func (keyboardApi *KeyboardAPI) GetKeyboardConnectors(c *gin.Context) {
	keyboardConnectors := keyboardApi.IKeyboardService.GetKeyboardConnectors()
	c.JSON(http.StatusOK, keyboardConnectors)
}

func (keyboardApi *KeyboardAPI) GetKeyTypes(c *gin.Context) {
	keyTypes := keyboardApi.IKeyboardService.GetKeyTypes()
	c.JSON(http.StatusOK, keyTypes)
}

func (keyboardApi *KeyboardAPI) Create(c *gin.Context) {
	var keyboardDTO dto.KeyboardDTO
	err := c.BindJSON(&keyboardDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	keyboard := mapper.ToKeyboard(keyboardDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	msg := keyboardApi.IKeyboardService.Create(keyboard)

	if msg == "" {
		c.JSON(http.StatusOK, "Keyboard added successfully.")
	} else {
		c.JSON(http.StatusBadRequest, msg)
	}
}

func (keyboardApi *KeyboardAPI) Update(c *gin.Context) {
	var keyboardDTO dto.KeyboardDTO
	err := c.BindJSON(&keyboardDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	msg := keyboardApi.IKeyboardService.Update(keyboardDTO)

	if msg == "" {
		c.JSON(http.StatusOK, "Keyboard updated successfully.")
	} else  {
		c.JSON(http.StatusBadRequest, msg)
	} 
}

func (keyboardApi *KeyboardAPI) Delete(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	error := keyboardApi.IKeyboardService.Delete(id)

	if error == nil {
		c.JSON(http.StatusOK, "Keyboard deleted successfully.")
	} else  {
		c.JSON(http.StatusBadRequest, error.Error())
	}
}

File: api-gateway/api/product_microservice/keyboard_api.py
import requests
from flask import request, jsonify
from utils import token_utils, role
from utils.routes.product_microservice import keyboard_api_routes


def get_all():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    r = requests.get(keyboard_api_routes.BASE + keyboard_api_routes.API +
                        keyboard_api_routes.GET_ALL
                        + "?page={}&pageSize={}".format(page, page_size))
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_number_of_records():
    r = requests.get(keyboard_api_routes.BASE + keyboard_api_routes.API +
                        keyboard_api_routes.GET_NUMBER_OF_RECORDS)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_by_id(id):
    r = requests.get(keyboard_api_routes.BASE + keyboard_api_routes.API +
                        keyboard_api_routes.GET_BY_ID + "/{}".format(id))
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def search_by_name():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    name = args.get("name")
    r = requests.get(keyboard_api_routes.BASE + keyboard_api_routes.API +
                        keyboard_api_routes.SEARCH_BY_NAME
                        + "?page={}&pageSize={}&name={}".format(page, page_size, name))
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_number_of_records_search():
    args = request.args.to_dict()
    name = args.get("name")
    r = requests.get(keyboard_api_routes.BASE + keyboard_api_routes.API +
                        keyboard_api_routes.GET_NUMBER_OF_RECORDS_SEARCH + "?name={}".format(name))
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def filter():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    data = request.json
    r = requests.post(keyboard_api_routes.BASE + keyboard_api_routes.API +
                         keyboard_api_routes.FILTER +
                         "?page={}&pageSize={}".format(page, page_size), json=data)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_number_of_records_filter():
    data = request.json
    r = requests.post(keyboard_api_routes.BASE + keyboard_api_routes.API +
                         keyboard_api_routes.GET_NUMBER_OF_RECORDS_FILTER, json=data)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_manufacturers():
    r = requests.get(keyboard_api_routes.BASE + keyboard_api_routes.API +
                        keyboard_api_routes.GET_MANUFACTURERS)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_keyboard_connectors():
    r = requests.get(keyboard_api_routes.BASE + keyboard_api_routes.API +
                        keyboard_api_routes.GET_KEYBOARD_CONNECTORS)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


def get_key_types():
    r = requests.get(keyboard_api_routes.BASE + keyboard_api_routes.API +
                        keyboard_api_routes.GET_KEY_TYPES)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def create():
    data = request.json
    headers = request.headers
    r = requests.post(keyboard_api_routes.BASE + keyboard_api_routes.API +
                         keyboard_api_routes.CREATE, json=data, headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def update():
    data = request.json
    headers = request.headers
    r = requests.put(keyboard_api_routes.BASE + keyboard_api_routes.API +
                        keyboard_api_routes.UPDATE, json=data, headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def delete():
    headers = request.headers
    r = requests.delete(keyboard_api_routes.BASE + keyboard_api_routes.API +
                           keyboard_api_routes.DELETE + "/{}".format(id), headers=headers)
    resp = jsonify(r.json())
    resp.status_code = r.status_code
    return resp
